datafusion/expr/src/accumulator.rs

```diff
@@ -44,3 +44,27 @@ pub trait Accumulator: Send + Sync + Debug {
     /// returns its value based on its current state.
     fn evaluate(&self) -> Result<ScalarValue>;
 }
+
+#[derive(Debug)]
+pub enum AggregateState {
+    Scalar(ScalarValue),
+    Array(ArrayRef),
+}
+
+impl AggregateState {
+    pub fn as_scalar(&self) -> Result<&ScalarValue> {
+        match &self {
+            Self::Scalar(v) => Ok(v),
+            _ => Err(DataFusionError::Execution(
+                "not a scalar aggregate".to_string(),
+            )),
```

Comment:
```suggestion
            _ => Err(DataFusionError::Internal(
                "AggregateState is not a scalar aggregate".to_string(),
            )),
```

I think marking this `Internal` might better signal to user if they hit this, they have seen some sort of bug with DataFusion